[MLIR][Python] Python binding support for AffineIfOp

kaitingwang opened the pull request:
Fix the AffineIfOp's default builder such that it takes in an IntegerSetAttr.   AffineIfOp has skipDefaultBuilders=1 which effectively skips the creation of the default AffineIfOp::builder on the C++ side. (AffineIfOp has two custom OpBuilder defined in the extraClassDeclaration.)  However, on the python side, _affine_ops_gen.py shows that the default builder is being created, but it does not accept IntegerSet and thus is useless.  This fix at line 411 makes the default python AffineIfOp builder take in an IntegerSet input and does not impact the C++ side of things.

kaitingwang committed:
[MLIR][Python] Python binding support for AffineIfOp

ftynse commented on mlir/python/mlir/dialects/affine.py:
```diff
@@ -156,3 +156,61 @@ def for_(
+        if cond_operands is None:
+            cond_operands = []
+
+        if not (actual_n_inputs := len(cond_operands)) == (
+            exp_n_inputs := cond.n_inputs
+        ):
```
Nit: I'm not sure using `:=` actually makes code shorter here, but less readable... Could we just assign variables above?

kaitingwang replied:
Got rid of the walrus assignment for readability.

ftynse commented on mlir/python/mlir/dialects/affine.py:
```diff
@@ -156,3 +156,61 @@ def for_(
+        results_: Optional[Type] = None,
+        *,
+        cond_operands: Optional[_VariadicResultValueT] = None,
+        hasElse: bool = False,
```
```suggestion
        has_else: bool = False,
```

kaitingwang replied:
Renamed.

ftynse commented on mlir/python/mlir/dialects/affine.py:
```diff
@@ -156,3 +156,61 @@ def for_(
+
+        super().__init__(results, cond_operands, cond)
+        self.regions[0].blocks.append(*[])
+        if hasElse:
```
```suggestion
        if has_else:
```

kaitingwang replied:
Renamed.

ftynse commented on mlir/python/mlir/dialects/affine.py:
```diff
@@ -156,3 +156,61 @@ def for_(
+            self.regions[1].blocks.append(*[])
+
+    @property
+    def then_block(self):
```
```suggestion
    def then_block(self) -> ir.Block:
```

kaitingwang replied:
Added the return type.

ftynse commented on mlir/python/mlir/dialects/affine.py:
```diff
@@ -156,3 +156,61 @@ def for_(
+        return self.regions[0].blocks[0]
+
+    @property
+    def else_block(self):
```
```suggestion
    def else_block(self) -> Optional[ir.Block]:
```

kaitingwang replied:
Added the Optional return type.

ftynse commented on mlir/python/mlir/dialects/affine.py:
```diff
@@ -156,3 +156,61 @@ def for_(
+    @property
+    def else_block(self):
+        """Returns the else block of the if operation."""
+        return self.regions[1].blocks[0]
```
Let's check that the region is not empty, and return `None` when it is. There's no guarantee that the `else` branch exists.

kaitingwang replied:
Added if statement to check for regions[1].block is empty, then return None.

ftynse reviewed (approved):
LGTM with comments addressed

kaitingwang commented:
@ftynse Much appreciated the reviewed!  Nice meeting you at the llvm developer conference last month!  In summary, I addressed all the points you raised: removed the walrus assignment for readability, renamed the `has_else` variable, checked for empty region[1].
